src/controllers/hid/hidio.cpp

```diff
@@ -0,0 +1,259 @@
+    Trace process("HidIO processInputReport");
+    unsigned char* pPreviousBuffer = m_pPollData[(m_pollingBufferIndex + 1) % kNumBuffers];
+    unsigned char* pCurrentBuffer = m_pPollData[m_pollingBufferIndex];
+    // Some controllers such as the Gemini GMX continuously send input reports even if it
+    // is identical to the previous send input report. If this loop processed all those redundant
+    // input report, it would be a big performance problem to run JS code for every  input report and
+    // would be unnecessary.
+    // This assumes that the redundant input report all use the same report ID. In practice we
+    // have not encountered any controllers that send redundant input report with different report
+    // IDs. If any such devices exist, this may be changed to use a separate buffer to store
+    // the last input report for each report ID.
+    if (bytesRead == m_lastPollSize &&
+            memcmp(pCurrentBuffer, pPreviousBuffer, bytesRead) == 0) {
+        return;
+    }
+    // Cycle between buffers so the memcmp above does not require deep copying to another buffer.
+    m_pollingBufferIndex = (m_pollingBufferIndex + 1) % kNumBuffers;
+    m_lastPollSize = bytesRead;
+    auto incomingData = QByteArray::fromRawData(
+            reinterpret_cast<char*>(pCurrentBuffer), bytesRead);
+
+    // Execute callback function in JavaScript mapping
+    // and print to stdout in case of --controllerDebug
+    emit receive(incomingData, mixxx::Time::elapsed());
+}
+
+QByteArray HidIo::getInputReport(unsigned int reportID) {
+    auto startOfHidGetInputReport = mixxx::Time::elapsed();
+    int bytesRead;
+
+    m_pPollData[m_pollingBufferIndex][0] = reportID;
+    // FIXME: implement upstream for hidraw backend on Linux
+    // https://github.com/libusb/hidapi/issues/259
+    bytesRead = hid_get_input_report(m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
+    qCDebug(m_logInput) << bytesRead << "bytes received by hid_get_input_report"
+                        << m_deviceInfo.formatName() << "serial #"
+                        << m_deviceInfo.serialNumber()
+                        << "(including one byte for the report ID:"
+                        << QString::number(static_cast<quint8>(reportID), 16)
+                                   .toUpper()
+                                   .rightJustified(2, QChar('0'))
+                        << ") - Needed: "
+                        << (mixxx::Time::elapsed() - startOfHidGetInputReport)
+                                   .formatMicrosWithUnit();
+
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes.
+        DEBUG_ASSERT(bytesRead <= kReportIdSize);
+        return QByteArray();
+    }
+
+    return QByteArray::fromRawData(
```

Comment:
Handing out a pointer to data that is managed by the thread in a public is unsafe, dangerous, and wrong.

Reply:
I suppose you could use the plain `QByteArray` constructor rather than `QByteArray::fromRawData` to make a copy? Ugly, but I don't have other ideas to ensure safety.

Reply:
Invoking this method from another thread via direct connection remains invalid, independent of the return type.